core/src/main/java/org/elasticsearch/index/search/MatchQuery.java

```diff
@@ -286,7 +286,11 @@ protected final Query termQuery(MappedFieldType fieldType, Object value, boolean
     }
 
     protected Query zeroTermsQuery() {
-        return zeroTermsQuery == DEFAULT_ZERO_TERMS_QUERY ? Queries.newMatchNoDocsQuery() : Queries.newMatchAllQuery();
+        if (zeroTermsQuery == DEFAULT_ZERO_TERMS_QUERY) {
+            return Queries.newMatchNoDocsQuery("No query supplied in match query");
```

Comment:
Nit: The zeroTermsQuery was left (or intentionally set) to the default here. This happens according to the docs when 'no query terms are available, e.g. after analysis removed them'. So maybe just mention sth. like "Matching no documents because no terms present"?